File: pcp/aufgaben/sw01/3-java-stack/Stack.java
package com.github.patrickbucher.pcp.stack;

public class Stack {

	private Element top = null;

	private int size = 0;

	public void push(Element e) {
		if (top == null) {
			top = e;
		} else {
			e.setNext(top);
			top = e;
		}
		size++;
	}

	public Element top() {
		if (isEmpty()) {
			return null;
		}
		return top;
	}

	public boolean pop() {
		if (top == null) {
			return false;
		}
		top = top.getNext();
		size--;
		return true;
	}

	public void print() {
		if (isEmpty()) {
			System.out.println("print - Stack is empty");
			return;
		}
		System.out.print("print - Stack contains: ");
		for (Element e = top; e != null; e = e.getNext()) {
			System.out.print(e.getValue() + ", ");
		}
		System.out.print("top element = " + top.getValue() + "\n");
	}

	public boolean isEmpty() {
		return size == 0;
	}

	public int size() {
		return size;
	}
}

File: pcp/aufgaben/sw01/2-stack/stack.c
#include <stdio.h>
#include <stdlib.h>
#include "stack.h"

// returns new empty stack
stack init() {
    stack s;
    s.index = STACK_EMPTY_INDEX;
    return s;
}

// adds element to stack
stack push(element e, stack s) {
    if (s.index + 1 < STACK_SIZE) {
        // there is space for one more element
        s.index++;
        s.stackArray[s.index] = e;
    } else {
        printf("ERROR - push: stack full! Cannot add %i\n", e);
    }
    return s;
}

// returns top element
element top(stack s) {
    if (s.index > STACK_EMPTY_INDEX) {
        return s.stackArray[s.index];
    } else {
        printf("ERROR - top: stack empty!\n");
        return STACK_DUMMY_ELEMENT;
    }
}

// removes top element from stack
stack pop(stack s) {
    if (s.index > STACK_EMPTY_INDEX) {
        s.index--;
    }
    return s;
}

// prints all elements
void print(stack s) {
    if (s.index > STACK_EMPTY_INDEX) {
        printf("print - Stack contains: ");
        int i;
        for (i = 0; i <= s.index; i++) {
            printf("%i, ", s.stackArray[i]);
        }
        printf("top element = %i\n", s.stackArray[s.index]);
    } else {
        printf("print - Stack is empty\n");
    }
}

int isEmpty(stack s) {
    if (s.index == STACK_EMPTY_INDEX) {
        return 1;
    } else {
        return 0;
    }
}

int size(stack s) {
    return s.index + 1;
}
